### DISCUSSION TITLE
Add Omnipod 5 IOB to xDrip

### DISCUSSION THREAD

**Discussion by gth001** [NightscoutFoundation/xDrip#2728] [Ideas]
This is a feature request to integrate with the Omnipod 5, to bring IOB, bolus, and carbs in from the Omnipod app.

This is a feature that the Gluroo app currently has. It would be cool if xDrip did the same.

How the function would work:
- User provides ISF, CR, target BG settings from Omnipod app to xDrip
- CGM provides BG
- User grants xDrip "Notification access" permission to read all notifications on the phone
- IOB = xDrip reads the IOB: 2.0u value from the Omnipod notification every ~5 minutes
- estimated bolus = any unusually large jump in IOB
- estimated carbs = CR*(bolus - (BG - target)/ISF))

Here is an example of an app that does this:
[Gluroo Omnipod 5 IOB integration documentation](https://gluroo.com/support/integration-omnipod-5/) (it's a bit outdated and doesn't mention the estimated carbs function they added yesterday)
<img width="739" alt="image" src="https://user-images.githubusercontent.com/10224533/224770246-a2c05c83-6023-4e12-a5e3-8c5f13869530.png">

**Comment by Navid200** [comment]
Can you provide screenshots of the Omnipod notifications?

**Reply by gdc27health** [reply]
I too use an Omnipod 5.  An IOB complication would be awesome to be able to show on a watch face.

Here is a screenshot of the Omnipod 5 Notification.  From a Galaxy S22 Ultra.  Sorry for the Weather and Dexcom notifications on the same screen

**Comment by gdc27health** [comment]
![Screenshot_20230601_130346_One UI Home](https://github.com/NightscoutFoundation/xDrip/assets/129467702/df7b8036-a645-4030-8d93-17eec0a8c672)

**Reply by Navid200** [reply]
@gdc27health Thanks

Would you also please expand that Omnipod notification and show any detail it may include?

**Reply by gdc27health** [reply]
See how the arrow on the right points up?  That means that this is expanded.  There is no other information.

**Comment by bart2** [comment]
I took a stab at implementing this feature: https://github.com/NightscoutFoundation/xDrip/pull/2928 (at least the IoB part).

**Comment by gjbadros** [comment]
Imitation is the most sincere form of flattery -- thank you! :-) (We're fans of XDrip, too!)

**Comment by gth001** [comment]
Is this feature working for anyone? It's not working for me.
I can scrape the IOB from the Omnipod notification using Tasker or using
Gluroo, but not with xDrip+.
Happy to troubleshoot/test. Just wondering if the feature is supposed to
work yet, since the toggle is available.

**Reply by bart2** [reply]
Lets try to troubleshoot this.

To double check, both "Enable IoB reporting in API" and "Fetch IoB from companion app" toggles are selected in "Inter-app settings" section?

Also, can you open chrome on your phone and go to "127.0.0.1:17580/pebble" and take a screenshot of the output?

**Reply by BillShepp** [reply]
I'm seeing IOB as part of the response from "127.0.0.1:17580/pebble". Note that I had to turn on "xDrip Web Service" from Inter-app settings.

Is there a way to expose the IOB data as part of the xDrip Complication Provider?

**Reply by bart2** [reply]
@BillShepp, I'm not familiar with xDrip Complication Provider. Can you explain how are you accessing it?

**Reply by gth001** [reply]
(Original question poster here)

Yes, I see the iob: number at 127.0.0.1:17580/pebble, and it's correct.

However, I don't see IOB on the xDrip main screen.

# What I currently see in xDrip  
(notice there's no IOB displayed, no green insulin markers/line, and the purple prediction line isn't taking IOB into account, and the "low predicted" alarm doesn't take IOB into account):
![image](https://github.com/NightscoutFoundation/xDrip/assets/10224533/93aa5e89-fd0b-452f-a526-2c9a3d77fe4c)

# What I expected  
(a screenshot when I was using Gluroo to scrape the IOB from Ominpod into xDrip. notice IOB is displayed, there are green insulin markers/line, and the purple prediction line takes IOB into account, and the "low predicted" alarm takes IOB into account):
![image](https://github.com/NightscoutFoundation/xDrip/assets/10224533/e8e78915-cd4f-459a-b169-a1f0bb5900ff)

**Comment by BillShepp** [comment]
I believe it’s a feature of the xDrip+ codebase (see
https://github.com/jamorham/xDrip-plus/blob/master/Documentation/WatchGuide.md).
With most watchfaces that allow configuring the complication I can choose
“xDrip Complication Provider” under the category “xDrip Prefs” from all the
complications available on the watch.

**Comment by gitanO0** [comment]
Is this working for folks?  I just got on an Omnipod 5, have granted the notification access, and enabled the setting but am not seeing any IOB numbers in xdrip.

I'm seeing the correct IOB on my fitbit so I guess the web service number is interacting correctly, just not the main graph? @bart2 ?

**Reply by gth001** [reply]
No, it never worked. I would like to share my disappointment with you. By that, I mean the feature description isn't at all what most people would interpret it to be. It should honestly be renamed to "standalone omnipod 5 to pebble/fitbit API pipeline" or something. Your Omnipod 5 IOB never actually goes into the multitude of IOB functions and variables in XDrip at all. Its only functionality is to create a new external variable for the Pebble smartwatch endpoint, so a smartwatch developer can pull the value to complications on smartwatches. However, apart from that, there is no functionality for users like you and me. 

If you would like to have your IOB actually pulled into xDrip, show up as mini-boluses in your graph, affect your prediction kinematics and soforth, someone DID implement this. It's called the Gluroo app. It will emulate a nightscout instance, send your IOB from the Omnipod 5 notification to it, and then xDrip will slurp that up from said nightscout instance. It works pretty well and hopefully xDrip implements it natively some day. But no rush, Gluroo+xDrip working in tandem works great for now. Super useful to diabetics on Omnipod.

**Reply by gitanO0** [reply]
will look into this, thx!

**Reply by gitanO0** [reply]
I've got this sorta kinda working.  I have the Gluroo app sending data to my nightsout site and I see bolus info as my IOB but do not see the basal doses or my full IOB (basal + bolus).  I'm pretty new to this so maybe I'm not getting the big picture here but I'd like to see in Nightscout the same IOB I'm seeing on the Omnipod 5 app and the Gluroo app.  

I also now have that bolus info going back to xdrip+ as well but again, would like to see all the insulin delivery info including the correct IOB and not just the bolus'. 

@gth001 is there something I can configure to do that?

Also, the Gluroo app on my Android phone is force closing a lot.  I've tried different things to see if I can fix that to no avail.

**Reply by gth001** [reply]
As for the force closing, I don't know, perhaps try setting the apps battery optimization to "unrestricted" in Android settings. Android might be trying to close it to save battery, but it needs to always run.

As for the IOB, at first it's inaccurate, but after 5+ hours it should start getting pretty close. 

The way it works is that it reads the Omnipod 5 notifacation every 5 mintues, IOB:1.25U, IOB:1.2U, IOB:1.15U waiting for it to increase. Then it sees IOB:1.65U. This causes it to log a 0.5U bolus into Nighscout (which xDrip picks up).

**Reply by Jinded** [reply]
> If you would like to have your IOB actually pulled into xDrip, show up as mini-boluses in your graph, affect your prediction kinematics and soforth, someone DID implement this. It's called the Gluroo app. It will emulate a nightscout instance, send your IOB from the Omnipod 5 notification to it, and then xDrip will slurp that up from said nightscout instance.
Hey if it's not too much trouble, would you mind helping me setup this functionality? I have xDrip+, Omnipod 5, and Gluroo on the same phone, and Gluroo is correctly grabbing IOB from the Omnipod notification, but I can't get this info to show up in xDrip+.

I have Gluroo Global Connect enabled, which I think is what simulates a Nightscout backend? From that screen I used Copy > xDrip+ to grab the URL that includes the API Secret, and pasted it into xDrip+ Settings > Cloud Upload > Nightscout Sync > Base URL. I also copied the API Secret into Inter-App Settings > xDrip Web Service Secret, and have tried toggling "Enable IoB reporting in API" and "Fetch IoB from Companion App" with the same result.

Gluroo reports receiving data in the Gluroo Global Connect setup screen, but I guess xDrip+ is not receiving any data back from it :(

**Comment by imarotte** [comment]
Im interested in connecting my Omnipod5 to xdrip as well but it sounds like it's not actually implemented yet. Are there any plans to add this in the future?

**Comment by gdc27health** [comment]
Use the GlucoDataHandler app available on the Google play store

He reads the xDrip data passes IOB and Acton more

**Reply by imarotte** [reply]
@gdc27health I appreciate the response, but I don't think this addresses any of the issues in this thread.
The issue is that xDrip does not get the IOB data from omnipod.

**Comment by gdc27health** [comment]
I understand.

Isaac I have identified an issue with the pebble endpoint and reported it
yesterday.

IOB not visible in pebble endpoint with current xDrip+ build #3795

**Comment by F242Thom** [comment]
Alternatively to getting this to work, is anyone working on creating an Omnipod 5 Widget and/or watch complications by scraping the notifications for IOB?  I confess GLUROO is not working right since the displays they use truncate the IOB so you only see one number instead of the sum...never mind why they have it split into a sum.  Also, Gluroo will invent carbs even if you don't list any for OP5.  Granted, Insulet has crap programmers so someone else has to take the point.

**Reply by gjbadros** [reply]
Next Gluroo WearOS release will have a separate complication for exactly what the IOB reports, and we anticipate a setting for turning off the carb estimation.  If you enter your ISF and CR into Gluroo accurately, the carb estimations work really well and they are de-duped with other carb entries if, for example, you used the AI vision feature to enter your meal.

**Comment by F242Thom** [comment]
Will they fix the iob to show one number instead of a sum?  I do not use their watch so I use their widget and get that on my watch via widget wear.  But the carb data is useless to me since I take semaglutide which slows digestion a bunch.  Also makes the IOB somewhat wrong as well since it radically changes ISF with moderate exercise like riding a bike.
As I said, a omnipod widget is all I need.

Yahoo Mail: Search, Organize, Conquer 
 
  On Tue, Aug 26, 2025 at 14:37, Greg ***@***.***> wrote:   


Next Gluroo WearOS release will have a separate complication for exactly what the IOB reports, and we anticipate a setting for turning off the carb estimation. If you enter your ISF and CR into Gluroo accurately, the carb estimations work really well and they are de-duped with other carb entries if, for example, you used the AI vision feature to enter your meal.

**Comment by F242Thom** [comment]
Can you remind me how to remove the xdrip widget from the phone lockscreen?   Ive been through the settings and can't seem to find it. Tia.

Yahoo Mail: Search, Organize, Conquer 
 
  On Tue, Aug 26, 2025 at 14:37, Greg ***@***.***> wrote:   


Next Gluroo WearOS release will have a separate complication for exactly what the IOB reports, and we anticipate a setting for turning off the carb estimation. If you enter your ISF and CR into Gluroo accurately, the carb estimations work really well and they are de-duped with other carb entries if, for example, you used the AI vision feature to enter your meal.

**Comment by BillShepp** [comment]
You usually remove a widget by pressing and holding on the widget.
“Remove” should appear at the top of the screen, drag it to “remove”.

**Comment by F242Thom** [comment]
Thanks.  

Yahoo Mail: Search, Organize, Conquer 
 
  On Tue, Aug 26, 2025 at 20:04, ***@***.***> wrote:   

 You usually remove a widget by pressing and holding on the widget.
“Remove” should appear at the top of the screen, drag it to “remove”.

**Comment by gdc27health** [comment]
Check out GlucoDataHandler

**Comment by F242Thom** [comment]
Thanks.  Have been using it and works well.   Scrapes iob from omnipod 5 notifications and doesn't invent carbs like gluroo and easier to use than xdrip.  Lots of new tools this decade. 
 Omnipod 5 probably could easily be made closed loop with a routine 2 hour use of the bolus calculator to prevent highs and routine use of activity feature for insulin sensitive folks who also exercise post prandial.  

Yahoo Mail: Search, Organize, Conquer 
 
  On Sat, Sep 20, 2025 at 13:23, ***@***.***> wrote:   

Check out GlucoDataHandler
